proof/refine/ARM/KHeap_R.thy

```diff
@@ -3921,4 +3935,315 @@ lemma threadGet_getObject:
   apply (simp add: threadGet_def threadRead_def oliftM_def getObject_def[symmetric])
   done
 
+lemma obj_at'_typ_at'[elim!]:
+  "obj_at' (P :: ('a :: pspace_storable) \<Rightarrow> bool) p s \<Longrightarrow>
+   obj_at' (\<top> :: ('a :: pspace_storable) \<Rightarrow> bool) p s"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+
+lemma sc_tcbs_of_pred_map_equiv:
+  "sc_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. scTCB x = Some t) p s = pred_map_eq t (scTCBs_of s) p"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def pred_map_eq_def pred_map_def
+                     projectKO_sc opt_map_def)
+
+lemma tcb_scs_of_pred_map_equiv:
+  "tcb_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. tcbSchedContext x = Some sc) p s = pred_map_eq sc (tcb_scs_of' s) p"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def pred_map_eq_def pred_map_def
+                     projectKO_tcb opt_map_def)
+
+lemma replySCs_of_pred_map_equiv:
+  "reply_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>a. replyNext a = Some (Head sc)) p s = pred_map_eq sc (replySCs_of s) p"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def pred_map_eq_def pred_map_def
+                     projectKO_tcb opt_map_def)
+
+lemma scReplies_of_pred_map_equiv:
+  "sc_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>a. scReply a = Some sc) p s = pred_map_eq sc (scReplies_of s) p"
+  by (clarsimp simp: obj_at'_real_def ko_wp_at'_def pred_map_eq_def pred_map_def
+                     projectKO_tcb opt_map_def)
+
+lemma tcbs_scs_sym_refs_equiv:
+  "tcbs_scs_sym_refs s \<Longrightarrow>
+   valid_objs' s \<Longrightarrow>
+   sc_at' p s \<or> tcb_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. scTCB x = Some t) p s = obj_at' (\<lambda>x. tcbSchedContext x = Some p) t s"
+  apply (rule iffI; subgoal_tac "sc_at' p s \<and> tcb_at' t s", clarsimp)
+     apply (clarsimp simp: sc_tcbs_of_pred_map_equiv tcb_scs_of_pred_map_equiv)
+     apply (clarsimp simp: sqheap_refs_inv_defs tcb_scs_of_pred_map_equiv)
+    apply (frule obj_at'_typ_at', simp)
+    apply (frule obj_at_ko_at', clarsimp)
+    apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+    apply (clarsimp simp: valid_sched_context'_def)
+   apply (clarsimp simp: sc_tcbs_of_pred_map_equiv tcb_scs_of_pred_map_equiv)
+   apply (clarsimp simp: sqheap_refs_inv_defs tcb_scs_of_pred_map_equiv)
+  apply (frule obj_at'_typ_at', simp)
+  apply (frule obj_at_ko_at', clarsimp)
+  apply (frule (1) tcb_ko_at_valid_objs_valid_tcb')
+  apply (clarsimp simp: valid_tcb'_def)
+  done
+
+lemma not_idle_scTCB:
+  "tcbs_scs_sym_refs s \<Longrightarrow>
+   valid_objs' s \<Longrightarrow>
+   valid_idle' s \<Longrightarrow>
+   p \<noteq> idle_sc_ptr \<Longrightarrow>
+   sc_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. scTCB x \<noteq> Some idle_thread_ptr) p s"
```

Comment:
These lemmas are really nice as well, I think there's been quite a few times where I've had to recreate something similar on the spot.